### PR TITLE
feat: emit error diagnostic when assign to import

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
### Description
1. Part of https://github.com/rolldown/rolldown/issues/2607
2. ref: https://github.com/evanw/esbuild/commit/394ff827ea819775ab54958e5c316df32a2be0aa#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR56

### CODE REVIEW

**Comment by IWANABETHATGUY** [comment]
* **#2703** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2703?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2701** <a href="https://app.graphite.dev/github/pr/rolldown/rolldown/2701?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @IWANABETHATGUY and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/rolldown/src/ast_scanner/import_assign_analyzer.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+use super::AstScanner;
+
+impl<'me, 'ast: 'me> AstScanner<'me, 'ast> {
+  pub fn check_import_assign(&mut self, ident: &IdentifierReference) {
+    if let Some(potential_update) =
+      self.potential_update_import_reference.get(&ident.reference_id())
+    {
+      match potential_update {
+        super::PotentialImportUpdateReference::Namespace => {
+          // TODO: namespace detection
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/import_assign_analyzer.rs]: assign to import with **namespace** is a little complex that needs some refactor, so I left in another pr.

File: crates/rolldown_testing/src/integration_test.rs [hunk 2]
```diff
@@ -156,18 +156,19 @@ impl IntegrationTest {
...
             "```text".into(),
             Cow::Owned(diagnostic.to_string()),
             "```".into(),
           ]
+          .concat()
         })
-        .collect::<Vec<_>>()
-        .join("\n");
+        .collect::<Vec<_>>();
+      rendered_diagnostics.sort();
```
Comment (IWANABETHATGUY) [inline crates/rolldown_testing/src/integration_test.rs]: make the diagnostic stable

File: crates/rolldown/src/ast_scanner/import_assign_analyzer.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            self.source.clone(),
+            ident.span,
+            ident.name.as_str().into(),
+          ));
+        }
+      }
+    }
+  }
+
+  #[allow(unused)]
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/ast_scanner/import_assign_analyzer.rs:29]: used for next pr

File: crates/rolldown/src/ast_scanner/impl_visit.rs [hunk 4]
```diff
@@ -144,6 +144,7 @@ impl<'me, 'ast: 'me> Visit<'ast> for AstScanner<'me, 'ast> {
         self.result.self_referenced_class_decl_symbol_ids.insert(symbol_id);
       }
     }
+    self.check_import_assign(ident);
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/impl_visit.rs]: Do you think what this PR does is similar to https://github.com/rolldown/rolldown/blob/bc16ada28c7122bf0246e8e3f46bb180179d6729/crates/rolldown/src/ast_scanner/impl_visit.rs#L183? Maybe we could handle them in one place?

File: crates/rolldown/src/ast_scanner/mod.rs [hunk 5]
```diff
@@ -83,6 +102,7 @@ pub struct AstScanner<'me, 'ast> {
   visit_path: Vec<AstKind<'ast>>,
   scope_stack: Vec<Option<ScopeId>>,
   options: Option<&'me SharedOptions>,
+  potential_update_import_reference: FxHashMap<ReferenceId, PotentialImportUpdateReference>,
```
Comment (hyf0) [inline crates/rolldown/src/ast_scanner/mod.rs]: Oxc semantic will record the import info, so there is method like `self.result.symbol_ref_db.get_flags(symbol_id).is_import()`.

Is replacing `is_const_var` with `is_import` what this PR want

https://github.com/rolldown/rolldown/blob/bc16ada28c7122bf0246e8e3f46bb180179d6729/crates/rolldown/src/ast_scanner/mod.rs#L626-L639

File: crates/rolldown/src/ast_scanner/impl_visit.rs (hunk 4 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/src/ast_scanner/impl_visit.rs]: Maybe they are not the same thing

Comment (IWANABETHATGUY) [reply crates/rolldown/src/ast_scanner/impl_visit.rs]: https://hyrious.me/esbuild-repl/?version=0.23.0&b=e%00entry.js%00import+*+as+ns+from+%27c%27%0Ans.test+%3D+1%0A%0Aconst+a+%3D+1%3B%0Aa.test+%3D+100%3B&b=%00dep.js%00export+const+dep+%3D+1000%3B&o=%7B%0A++treeShaking%3A+true%2C%0A++external%3A+%5B%22c%22%2C+%22a%22%2C+%22b%22%5D%2C%0A%22bundle%22%3A+true%2C%0Aformat%3A+%22esm%22%2C%0Aplatform%3A+%22node%22%2C%0Aminify%3A+false%0A%7D

Comment (IWANABETHATGUY) [reply crates/rolldown/src/ast_scanner/impl_visit.rs]: here is an example, forbid_const_assign don't care object property assgin, but for a `ns` import it needs to be reported.

Comment (hyf0) [reply crates/rolldown/src/ast_scanner/impl_visit.rs]: Seems just a special pattern of forbidding assign.

File: crates/rolldown/tests/esbuild/default/assign_to_import/artifacts.snap [hunk 6]
```diff
@@ -2,110 +2,113 @@
...
+   │                           ┬  
+   │                           ╰── Imports are immutable in JavaScript. To modify the value of this import, you must export a setter function in the imported file (e.g. 'setX') and then import and call that function here instead.
+───╯
+
+```
+## ASSIGN_TO_IMPORT
+
+```text
+[ASSIGN_TO_IMPORT] Error: Cannot assign to import 'x'
+   ╭─[good0.js:1:28]
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/default/assign_to_import/artifacts.snap:95]: Maybe false positive case in oxc, https://oxc-project.github.io/oxc/playground/?code=3YCAAICigICAgICAgIC0G8rnONK89ITJ3zrK%2FUP5L%2FiZLeiyawJpf0w9TqWLqZyk6ZSj338lRQCA

File: crates/rolldown/src/ast_scanner/mod.rs (hunk 5 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/src/ast_scanner/mod.rs]: Nice catch.

**Review by hyf0 — APPROVED** [review]

**Pull request merged**